oauth2client/dictionary_storage.py

```diff
@@ -0,0 +1,75 @@
+# Copyright 2014 Google Inc. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Dictionary storage for OAuth2 Credentials."""
+
+from oauth2client.client import OAuth2Credentials
+from oauth2client.locked_storage import LockedStorage
+
+
+class DictionaryStorage(LockedStorage):
+    """Store and retrieve credentials to and from a dictionary-like object."""
+
+    def __init__(self, dictionary, key, lock=None):
+        """Construct a DictionaryStorage instance.
+
+        Args:
+            dictionary: A dictionary or dictionary-like object.
+            key: A hashable or a function returning a hashable. The credentials
+                 will be stored in ``dictionary[key]``.
+            lock: An optional threading.Lock-like object. The lock will be
+                  acquired before anything is written or read from the
+                  dictionary.
+        """
+        super(DictionaryStorage, self).__init__(lock=lock)
+        self._dictionary = dictionary
+        self._key = key
+
+    def _get_key(self):
+        """Return the key for storing the credentials.
+
+        If self._key is a callable, it will return the result of calling
+        self._key.
+        """
+        if callable(self._key):
+            return self._key()
+        else:
+            return self._key
+
+    def locked_get(self):
+        """Retrieve the credentials from the dictionary, if they exist."""
+        key = self._get_key()
+        serialized = self._dictionary.get(key)
+
+        if serialized is None:
+            return None
+
+        credentials = OAuth2Credentials.from_json(serialized)
+        credentials.set_store(self)
+
+        return credentials
+
+    def locked_put(self, credentials):
+        """Save the credentials to the dictionary."""
+        key = self._get_key()
+        serialized = credentials.to_json()
+        self._dictionary[key] = serialized
+
+    def locked_delete(self):
+        """Remove the credentials from the dictionary, if they exist."""
+        key = self._get_key()
+        try:
```

Comment:
For the case of ensuring the absence of a key in a dictionary when you don't know going in whether or not the key is present in the dictionary: `my_dictionary.pop(my_key, None)`, rather than a try/except.